### ISSUE TITLE
flux bootstrap assumes port 22 when cloning?

### ISSUE THREAD

**Issue by jonaskello** [fluxcd/flux2#642]
I tried to use `flux bootstrap gitlab` on a hosted gitlab server.

```
flux bootstrap gitlab --owner=xxx --repository=fleet-infra --hostname=gitlab.myhost.com --path=my-cluster --branch=master --kubeconfig /{redacted}/kube_config_cluster.yml
```

The first time it failed like this:

```
► applying sync manifests
◎ waiting for cluster sync
✗ unable to clone 'ssh://{redacted}/fleet-infra', error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```

Then I thought I just try again with the same command if the error was some temporary thing. Since the command should be idempotent I would expect this to work. This time it fails like this instead:

```
► generating sync manifests
► applying sync manifests
◎ waiting for cluster sync
✗ unable to clone 'ssh://{redacted}/fleet-infra', error: dial tcp 10.xx.xx.x:22: connect: connection timed out
```

Now there are two things that `flux bootstrap gitlab` seems to get wrong the second time and I'm not sure why. The first is the DNS, it does not resolve to that IP if I use try nslookup. Not sure why. However that is not the main issue.

The main issue is that we have a non-standard ssh port for our gitlab server. However it seems `flux bootstrap gitlab` assumes port `22`. So even if I could get the DNS issue fixed it would still not work becuase of the invalid port. Is `flux` finding the correct port somewhere from gitlab or does it just assume `22`? I would argue that the correct way would be to check the clone url in the gitlab api but perhaps it is already done this way?

It is also a bit strange that the second run gives a different error than the first run. I ran these commands in the same terminal without any environmental changes that I know of.

**Comment by jonaskello** [comment]
I think I might understand why the command sometimes gives the auth error instead. The same hostname that the gitlab server is hosted on has another ssh service (not gitlab) running on port 22. Probably `flux` is hitting that service instead of gitlab and therefore getting auth error. So the root cause is probbably that we have a non-standard port but `flux`  assumes 22.

**Comment by jonaskello** [comment]
Ok, so digging deeper I found the `--ssh-hostname` parameter which actually is not just a hostname it can also take a port. So I changed my command to include this option:

```bash
flux bootstrap gitlab --owner=xxx --repository=fleet-infra --hostname=gitlab.myhost.com --path=my-cluster --branch=master --kubeconfig /{redacted}/kube_config_cluster.yml  --ssh-hostname gitlab.myhost.com:1234
```

Now I got this error instead:

```
► generating sync manifests
► applying sync manifests
◎ waiting for cluster sync
✗ unable to clone 'ssh://git@gitlab.divid.se:2201/divid-it/fleet-infra', error: ssh: handshake failed: knownhosts: key is unknown
```

So digging even deeper I found #475 and realised this is not from the clone command but from within the cluster. So probably the first time I ran it flux decided the ssh-host name and kept it at port 22. Now that I specified a correct port it did not update that so that is why the hostname was invalid (or actually just the port part was invalid).

The solution was to run `flux uninstall` to remove everything and then re-run `flux bootstrap` with correct `--ssh-host' option from the start. Then it finally worked :-).
